pkg/flow/internal/controller/node_with_dependants.go

```diff
@@ -0,0 +1,21 @@
+package controller
+
+import (
+	"time"
+
+	"github.com/grafana/agent/component"
+)
+
+// NodeWithDependants must be implemented by BlockNode that can trigger other nodes to be evaluated.
+type NodeWithDependants interface {
+	BlockNode
+
+	// LastUpdateTime returns the time corresponding to the last time where the node changed its exports.
+	LastUpdateTime() time.Time
+
+	// Exports returns the current set of exports from the managed component.
+	Exports() component.Exports
+
+	// ID returns the component ID of the managed component from its River block.
+	ID() ComponentID
+}
```

Comment:
I think there's a few issues with this interface:

* The name is potentially misleading. The name implies it's about a characteristic of a particular node, but the description is more about whether a BlockNode can emit an event informing its dependents to be updated. In fact, all nodes can have dependents regardless of whether they implement this interface. 

* The set of methods is perhaps too broad; it's not obvious why a node which can emit updates to cause other nodes to be evaluated _needs_ to have component exports, or a component ID.

Because of the points above, I think it's hard to justify introducing this interface as a new concept to keep inside one's mental model of the controller code.  

I would recommend that we aim to _not_ introduce this interface, opting instead for the following:

1. Move `LastUpdateTime()` to BlockNode instead. This would be a reasonable expectation of a BlockNode; as all BlockNodes support evaluating, it makes sense for them to also report their last update time. (Other methods should not be moved, just LastUpdateTime.)
2. Instead of changing code to accept `NodeWithDependents`, update it to accept a `BlockNode` instead. You will need to add some type switching when a more concrete type is necessary.
    * This will also have the added benefit of making some of the changed code a bit shorter; `OnNodeWithDependantsUpdate` is a bit harder to read than `OnBlockNodeUpdate`. 

WDYT?

Reply:
Makes sense, I was not really happy with the name in the first place so I'm happy to change this :)

Reply:
I made the changes to see how it looks but after putting more thought into it I'm not 100% happy with it:
- For the component node, LastUpdateTime actually refers to the last time its exports were updated, which is not necessarily tied to when the node was updated or evaluated.
- In the Evaluate dependant I use a type cast to cache the exports because the BlockNode does not have the Export() function but I feel like this should be the default behavior. All BlockNodes can be evaluated which means that they have a concept of "Arguments" (inputs). I think that it actually makes sense that they also have a concept of "Exports" (outputs) and that these exports should be cached if not nil.

But for now the definitions of Arguments and Exports are really tied to components and not nodes so it would require bigger changes.

Reply:
> For the component node, LastUpdateTime actually refers to the last time its exports were updated, which is not necessarily tied to when the node was updated or evaluated.

Ah, I see. Honestly, I don't think a node should have to track its own export time anyway; the controller could be tracking that when `OnBlockNodeUpdate` is called. What do you think? Should that be scoped into this PR so we can take it out of the interface?

Reply:
Sounds good, the change is small and the time is only used by one metric. I think that it is worth doing it in the PR as it prevents introducing redundant code. I will take care of it